src/rgw/rgw_rest_conn.h

```diff
@@ -67,8 +67,12 @@ inline param_vec_t make_param_list(const std::map<std::string, std::string> *pp)
 
 class RGWRESTConn
 {
+  /* the endpoint is not able to connect if the timestamp is not real_clock::zero */
+  using endpoint_status_map = std::unordered_map<std::string, std::atomic<ceph::real_time>>;
```

Comment:
This is a comment @cbodley made in [this](https://github.com/ceph/ceph/pull/52812#discussion_r1296188370) PR on the same changes, which have been moved and updated here.
> `endpoints_status` could probably just be a `std::map<std::string, ceph::real_time>` that only stores the failed endpoints. `std::unordered_map` is probably overkill when we only expect a handful of endpoints per zone. `boost::container::flat_map` could be a good alternative to reduce memory allocations

Here are the reasons why I still choose to use the `std::unordered_map` to maintain the status for each and every endpoint are the following:
- the `endpoints_status` map itself is created when the RGWRESTConn is instantiated and won't be updated after that, hence the map structure itself don't need to be thread-safe.
- however the timestamp/status of each endpoint might be accessed simultaneously by multiple threads so it should be handled in a thread-safe way. In order to reduce the lock granularity, I choose to use `std::atomic` for each endpoint instead of locking the entire map. Since majority operations on the map would be read-only, this can reduce the overall time spent on locking.
- `std::atomic` is neither copyable nor movable hence I'm not able to use `boost::container::flat_map` here.